### PR TITLE
Eventlog java

### PR DESCRIPTION
**Pull request by kc1212**
Implement java API for event logging. For now we aim to just implement three API - init, log and get. The tests don't pass because there's a bug in the eventlog branch.

### CODE REVIEW

File: eventlog/service.go [hunk 1]
```diff
@@ -81,9 +86,43 @@ func (s *Service) Log(req *LogRequest) (*LogResponse, error) {
...
+	return &GetEventResponse{
+		Event: e,
+	}, nil
+}
+
 const contractName = "eventlog"
 
-func (s *Service) decodeAndCheckEvent(eventBuf []byte) (*Event, error) {
+func (s *Service) decodeAndCheckEvent(coll collection.Collection, eventBuf []byte) (*Event, error) {
 	// Check the timestamp of the event: it should never be in the future,
```
Comment (ineiti) [inline eventlog/service.go]: The timestamp _might_ be in the future, because the node's clock is running too slow! So I think we really need to accept a time-window here that is independent of the `blockinterval`. This should probably be a configuration parameter of the event-logger. Something like 10 minutes or so.

The alternative is if the client-instruction creates the timestamp itself, then we could be a bit more rigid to what timestamps we accept.

File: external/java/pom.xml [hunk 2]
```diff
@@ -54,7 +54,7 @@
         <dependency>
             <groupId>com.google.protobuf</groupId>
             <artifactId>protobuf-java</artifactId>
-            <version>3.4.0</version>
+            <version>3.5.1</version>
```
Comment (ineiti) [inline external/java/pom.xml]: Out of curiosity: why did you need this new version here?

File: external/java/src/main/java/ch/epfl/dedis/lib/darc/Darc.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class Darc {
```
Comment (ineiti) [inline external/java/src/main/java/ch/epfl/dedis/lib/darc/Darc.java]: This needs some little documentation of what the `Darc` class is. Also for each method, please add a small
```
/**
 * description
 * params:
 */
```

File: external/java/src/main/java/ch/epfl/dedis/lib/darc/IdentityDarc.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package ch.epfl.dedis.lib.darc;
+
+import ch.epfl.dedis.lib.exception.CothorityCryptoException;
+import ch.epfl.dedis.proto.DarcProto;
+import com.google.protobuf.ByteString;
+
+import javax.xml.bind.DatatypeConverter;
+
+public class IdentityDarc implements Identity {
```
Comment (ineiti) [inline external/java/src/main/java/ch/epfl/dedis/lib/darc/IdentityDarc.java]: Is this directly imported from `cothority`? If yes, you can leave it like that. If you wrote it, please add a class-comment here.

**Review by ineiti — CHANGES_REQUESTED** [review]
Please add class-comments in java. 2 sentences are enough for now.

Also add method-comments using the
```
/**
 * description
 * @param
 * @return
 */
```
format.

File: eventlog/service.go (hunk 1 above)
Comment (kc1212) [reply eventlog/service.go]: Your suggestion sounds reasonable, I'll note it down and make an issue on cothority when we merge.

File: external/java/pom.xml (hunk 2 above)
Comment (kc1212) [reply external/java/pom.xml]: I think this is a side-effect of debugging protobuf issues. I can change it back, if you'd like.

Comment (ineiti) [reply external/java/pom.xml]: No - I was just wondering if there is a bug in 3.4.0 that you wanted to avoid. If possible, I think the latest version should be best. So keep it.

File: eventlog/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# EventLog
```
Comment (ineiti) [inline eventlog/README.md]: Wonderful!

File: external/java/src/main/java/ch/epfl/dedis/lib/eventlog/Event.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * An instance of an Event can be sent and stored by the EventLog service.
+ */
+public class Event {
+    private long when; // in nano seconds
+    private String topic;
+    private String content;
+
+    public Event(long when, String topic, String content) {
```
Comment (ineiti) [inline external/java/src/main/java/ch/epfl/dedis/lib/eventlog/Event.java]: This really needs some documentation - at least to tell that `long` is in nanoseconds or so.

**Comment by ineiti** [comment]
Please add at least documentation for every method for the classes you wrote yourself and that have not been written by somebody else. This includes `ClientTransaction`, `Instruction` (and all dependencies such as `Spawn`, `Invoke`, `Delete`), `Argument`, `Event`, `EventLog`.

**Comment by kc1212** [comment]
I thought I had documentation for all of these. Perhaps I forgot to push, I'll check when I'm in the office.

On 5 June 2018 07:38:58 CEST, Linus Gasser <notifications@github.com> wrote:
>Please add at least documentation for every method for the classes you
>wrote yourself and that have not been written by somebody else. This
>includes `ClientTransaction`, `Instruction` (and all dependencies such
>as `Spawn`, `Invoke`, `Delete`), `Argument`, `Event`, `EventLog`.
>
>-- 
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/dedis/student_18_omniledger/pull/69#issuecomment-394587824

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Review by ineiti — APPROVED** [review]
